File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        hash_map = {}  # To store the last index of each element
        
        for index, val in enumerate(nums):
            if val in hash_map and index - hash_map[val] <= k:
                return True  # Found a valid duplicate within the range
            hash_map[val] = index  # Update the last index of the element
        
        return False  # No valid duplicate found

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        int left=0;
        HashSet<Integer> lookUp=new HashSet<>();

        for(int right=0;right<nums.length;right++){
            if(lookUp.size()>k){
                lookUp.remove(nums[left]);
                left++;
            }
            if(lookUp.contains(nums[right]) && lookUp.contains(nums[left]) && lookUp.size()<=k){
                return true;
            }
            lookUp.add(nums[right]);
        }
        return false;
    }
}